File: Baekjoon/트리/9934 - 완전 이진 트리.cpp
#include<vector>
#include<iostream>
#include<math.h>

using namespace std;

int k;
vector<int> tmp;
vector<int> result[11];

void finding(int s, int e, int d)
{
	if (s == e) {
		result[d].push_back(tmp[s]);
		return;
	}
	
	// 완전 이진 트리, 중위 순회, 가운 데 있는 값 root
	int mid = (s + e) / 2;
	result[d].push_back(tmp[mid]);

	// root 제외 좌우 나누어 탐색.
	finding(s, mid - 1, d + 1);
	finding(mid + 1, e, d + 1);
}

int main()
{
	cin >> k;

	for (int i = 0; i < pow(2, k) - 1; i++) {
		int num;
		cin >> num;
		tmp.push_back(num);
	}

	finding(0, tmp.size() - 1, 0);

	for (int i = 0; i < k; i++)
	{
		for (int j = 0; j < result[i].size(); j++)
			cout << result[i][j] << ' ';
		cout << endl;		
	}
}

File: Baekjoon/9934 - 완전 이진 트리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
	static StringBuilder sb = new StringBuilder();
	static ArrayList[] result;
	static ArrayList tmp;
	public static void main(String args[]) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = (int)Math.pow(Integer.parseInt(br.readLine()), 2) - 1;
		tmp = new ArrayList<>();
		result = new ArrayList[n];
		for(int i = 0; i < n; i++)
			result[i] = new ArrayList<>();
		
		
		String[] s = br.readLine().split(" ");
		
		for(String ss : s)
			tmp.add(ss);
		
		
		find(0, tmp.size() - 1, 0);
		
		for(int i = 0; i < result.length; i++)
		{
			for(int j = 0; j < result[i].size(); j++)
				sb.append(result[i].get(j) + " ");
			sb.append("\n");
		}	
		System.out.println(sb);
	}
	static private void find(int s, int e, int d) {
		if(s==e) {
			result[d].add(tmp.get(s));
			return;
		}
		int mid = (s + e) / 2;
		result[d].add(tmp.get(mid));
		find(s, mid - 1, d + 1);
		find(mid + 1, e, d + 1);
			
	}
}